File: lohas-papers-backend/app/services/pdf_extractor.py
"""PDF text extraction and section splitting for academic papers."""

import io
import logging
import re

import httpx
from pypdf import PdfReader

logger = logging.getLogger(__name__)

MAX_PDF_SIZE = 20 * 1024 * 1024  # 20 MB
DOWNLOAD_TIMEOUT = 30.0

# Common section headers in academic papers
_SECTION_PATTERNS = [
    r"(?i)^(abstract)\s*$",
    r"(?i)^(introduction)\s*$",
    r"(?i)^(background)\s*$",
    r"(?i)^(methods?|materials?\s+and\s+methods?|experimental\s+methods?)\s*$",
    r"(?i)^(results?)\s*$",
    r"(?i)^(results?\s+and\s+discussion)\s*$",
    r"(?i)^(discussion)\s*$",
    r"(?i)^(conclusion|conclusions|concluding\s+remarks?)\s*$",
    r"(?i)^(limitations?)\s*$",
    r"(?i)^(acknowledgements?|acknowledgments?)\s*$",
    r"(?i)^(references|bibliography)\s*$",
    r"(?i)^(supplementary|supporting\s+information)\s*",
    # Numbered sections: "1. Introduction", "2 Methods", "I. Introduction"
    r"(?i)^(?:\d+\.?\s+|[IVX]+\.?\s+)(introduction|background|methods?|materials?\s+and\s+methods?|results?|discussion|conclusion|conclusions|limitations?|acknowledgements?|references)",
]

# Sections to exclude from translation
_EXCLUDE_SECTIONS = {"references", "bibliography", "acknowledgements", "acknowledgments", "supplementary", "supporting information"}


async def extract_text_from_url(pdf_url: str) -> str:
    """Download a PDF from URL and extract its text content.

    Returns the full text as a single string.
    Raises ValueError if the PDF is too large or download fails.
    """
    try:
        headers = {
            "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
            "Accept": "application/pdf,*/*",
        }
        async with httpx.AsyncClient(
            timeout=DOWNLOAD_TIMEOUT,
            follow_redirects=True,
        ) as client:
            resp = await client.get(pdf_url, headers=headers)
            resp.raise_for_status()

            content_length = len(resp.content)
            if content_length > MAX_PDF_SIZE:
                raise ValueError(
                    f"PDF too large: {content_length / 1024 / 1024:.1f} MB (max {MAX_PDF_SIZE / 1024 / 1024:.0f} MB)"
                )

            pdf_bytes = io.BytesIO(resp.content)

    except httpx.HTTPStatusError as e:
        raise ValueError(f"PDF download failed: HTTP {e.response.status_code}") from e
    except httpx.TimeoutException:
        raise ValueError("PDF download timed out") from None
    except ValueError:
        raise
    except Exception as e:
        raise ValueError(f"PDF download failed: {e}") from e

    # Extract text from PDF
    try:
        reader = PdfReader(pdf_bytes)
        pages_text = []
        for page in reader.pages:
            text = page.extract_text()
            if text:
                pages_text.append(text)

        full_text = "\n\n".join(pages_text)

        if not full_text.strip():
            raise ValueError("No text could be extracted from the PDF (may be image-based)")

        logger.info("Extracted %d characters from %d pages", len(full_text), len(reader.pages))
        return full_text

    except ValueError:
        raise
    except Exception as e:
        raise ValueError(f"PDF text extraction failed: {e}") from e


def split_into_sections(text: str) -> list[dict]:
    """Split extracted PDF text into logical sections.

    Returns a list of dicts: [{"name": "Introduction", "text": "..."}]
    If no section headers are detected, returns the entire text as one section.
    """
    lines = text.split("\n")
    sections: list[dict] = []
    current_name = "Full Text"
    current_lines: list[str] = []

    for line in lines:
        stripped = line.strip()
        if not stripped:
            current_lines.append("")
            continue

        # Check if this line is a section header
        matched_section = _match_section_header(stripped)
        if matched_section:
            # Save previous section
            if current_lines:
                section_text = "\n".join(current_lines).strip()
                if section_text:
                    sections.append({"name": current_name, "text": section_text})
            current_name = matched_section
            current_lines = []
        else:
            current_lines.append(line)

    # Save last section
    if current_lines:
        section_text = "\n".join(current_lines).strip()
        if section_text:
            sections.append({"name": current_name, "text": section_text})

    # If no sections were detected (only "Full Text"), return as-is
    if len(sections) <= 1:
        return [{"name": "Full Text", "text": text.strip()}]

    # Filter out unwanted sections (references, acknowledgements, etc.)
    filtered = []
    for s in sections:
        name_lower = s["name"].lower()
        if name_lower in _EXCLUDE_SECTIONS:
            continue
        # Skip very short sections (likely noise)
        if len(s["text"]) < 50:
            continue
        filtered.append(s)

    return filtered if filtered else [{"name": "Full Text", "text": text.strip()}]


def _match_section_header(line: str) -> str | None:
    """Check if a line matches a known section header pattern.

    Returns the canonical section name or None.
    """
    for pattern in _SECTION_PATTERNS:
        m = re.match(pattern, line)
        if m:
            # Extract the section name from the first capture group
            raw_name = m.group(1) if m.lastindex else line
            return _normalize_section_name(raw_name)
    return None


def _normalize_section_name(name: str) -> str:
    """Normalize section name to a canonical form."""
    name_lower = name.lower().strip()

    mapping = {
        "abstract": "Abstract",
        "introduction": "Introduction",
        "background": "Background",
        "method": "Methods",
        "methods": "Methods",
        "materials and methods": "Methods",
        "materials & methods": "Methods",
        "experimental methods": "Methods",
        "result": "Results",
        "results": "Results",
        "results and discussion": "Results and Discussion",
        "discussion": "Discussion",
        "conclusion": "Conclusion",
        "conclusions": "Conclusion",
        "concluding remarks": "Conclusion",
        "limitation": "Limitations",
        "limitations": "Limitations",
        "acknowledgement": "Acknowledgements",
        "acknowledgements": "Acknowledgements",
        "acknowledgment": "Acknowledgements",
        "acknowledgments": "Acknowledgements",
        "references": "References",
        "bibliography": "References",
        "supplementary": "Supplementary",
        "supporting information": "Supplementary",
    }

    return mapping.get(name_lower, name.title())

File: lohas-papers-web/src/lib/backend/pdf-extractor.ts
/**
 * PDF text extraction and section splitting for academic papers.
 *
 * Note: In the Python version, pypdf was used for PDF parsing.
 * In this TypeScript/Vercel version, we use pdf-parse (or raw text extraction).
 * Since pdf-parse is a heavy dependency, we'll do a simpler approach:
 * fetch the PDF as arraybuffer and use a lightweight parser.
 *
 * For Vercel deployment, we avoid native dependencies.
 * We use the pdf.js-based extraction via dynamic import.
 */

import { lookup } from "node:dns/promises";
import { isIP } from "node:net";

const MAX_PDF_SIZE = 20 * 1024 * 1024; // 20 MB
const DOWNLOAD_TIMEOUT_MS = 30_000;
const MAX_REDIRECTS = 3;

// Common section headers in academic papers
const SECTION_PATTERNS: RegExp[] = [
  /^(abstract)\s*$/i,
  /^(introduction)\s*$/i,
  /^(background)\s*$/i,
  /^(methods?|materials?\s+and\s+methods?|experimental\s+methods?)\s*$/i,
  /^(results?)\s*$/i,
  /^(results?\s+and\s+discussion)\s*$/i,
  /^(discussion)\s*$/i,
  /^(conclusion|conclusions|concluding\s+remarks?)\s*$/i,
  /^(limitations?)\s*$/i,
  /^(acknowledgements?|acknowledgments?)\s*$/i,
  /^(references|bibliography)\s*$/i,
  /^(supplementary|supporting\s+information)\s*/i,
  // Numbered sections
  /^(?:\d+\.?\s+|[IVX]+\.?\s+)(introduction|background|methods?|materials?\s+and\s+methods?|results?|discussion|conclusion|conclusions|limitations?|acknowledgements?|references)/i,
];

// Sections to exclude from translation
const EXCLUDE_SECTIONS = new Set([
  "references",
  "bibliography",
  "acknowledgements",
  "acknowledgments",
  "supplementary",
  "supporting information",
]);

export async function extractTextFromUrl(pdfUrl: string): Promise<string> {
  let pdfBuffer: ArrayBuffer;

  try {
    const resp = await fetchPdfSafely(pdfUrl);

    if (!resp.ok) {
      throw new Error(`PDF download failed: HTTP ${resp.status}`);
    }

    const contentType = resp.headers.get("content-type")?.toLowerCase() ?? "";
    if (
      contentType &&
      !contentType.includes("application/pdf") &&
      !contentType.includes("application/octet-stream") &&
      !contentType.includes("binary/octet-stream")
    ) {
      throw new Error(`Unexpected PDF content type: ${contentType}`);
    }

    pdfBuffer = await readLimitedArrayBuffer(resp);
  } catch (err) {
    if (err instanceof DOMException && err.name === "TimeoutError") {
      throw new Error("PDF download timed out");
    }
    throw err instanceof Error
      ? err
      : new Error(`PDF download failed: ${err}`);
  }

  // Extract text using pdf-parse (lazy import)
  try {
    // Dynamic import to avoid bundling issues
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const pdfParse = require("pdf-parse") as (
      buffer: Buffer,
    ) => Promise<{ text: string; numpages: number }>;

    const buffer = Buffer.from(pdfBuffer);
    const result = await pdfParse(buffer);

    const fullText = result.text;
    if (!fullText.trim()) {
      throw new Error(
        "No text could be extracted from the PDF (may be image-based)",
      );
    }

    console.info(
      `Extracted ${fullText.length} characters from ${result.numpages} pages`,
    );
    return fullText;
  } catch (err) {
    if (err instanceof Error && err.message.includes("No text")) {
      throw err;
    }
    throw new Error(`PDF text extraction failed: ${err}`);
  }
}

async function fetchPdfSafely(
  rawUrl: string,
  redirectCount: number = 0,
): Promise<Response> {
  const url = await validatePublicHttpUrl(rawUrl);
  const resp = await fetch(url, {
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
      Accept: "application/pdf,*/*",
    },
    redirect: "manual",
    signal: AbortSignal.timeout(DOWNLOAD_TIMEOUT_MS),
  });

  if (isRedirect(resp.status)) {
    if (redirectCount >= MAX_REDIRECTS) {
      throw new Error("PDF download redirected too many times");
    }

    const location = resp.headers.get("location");
    if (!location) {
      throw new Error("PDF download redirect missing Location header");
    }

    return fetchPdfSafely(new URL(location, url).toString(), redirectCount + 1);
  }

  return resp;
}

async function validatePublicHttpUrl(rawUrl: string): Promise<string> {
  let url: URL;
  try {
    url = new URL(rawUrl);
  } catch {
    throw new Error("Invalid PDF URL");
  }

  if (url.protocol !== "https:" && url.protocol !== "http:") {
    throw new Error("PDF URL must use http or https");
  }

  if (url.username || url.password) {
    throw new Error("PDF URL must not contain credentials");
  }

  await assertPublicHostname(url.hostname);
  return url.toString();
}

async function assertPublicHostname(hostname: string): Promise<void> {
  const normalized = hostname.toLowerCase();
  if (normalized === "localhost" || normalized.endsWith(".local")) {
    throw new Error("PDF URL hostname is not allowed");
  }

  const records = await lookup(hostname, { all: true, verbatim: true });
  if (records.length === 0) {
    throw new Error("PDF URL hostname could not be resolved");
  }

  if (records.some((record) => isBlockedAddress(record.address))) {
    throw new Error("PDF URL resolves to a private or reserved address");
  }
}

function isBlockedAddress(address: string): boolean {
  const version = isIP(address);
  if (version === 4) return isBlockedIPv4(address);
  if (version === 6) return isBlockedIPv6(address);
  return true;
}

function isBlockedIPv4(address: string): boolean {
  const [a, b, c] = address.split(".").map((part) => Number(part));
  if (!Number.isInteger(a) || !Number.isInteger(b) || !Number.isInteger(c)) {
    return true;
  }

  return (
    a === 0 ||
    a === 10 ||
    a === 127 ||
    (a === 169 && b === 254) ||
    (a === 172 && b >= 16 && b <= 31) ||
    (a === 192 && b === 168) ||
    (a === 100 && b >= 64 && b <= 127) ||
    (a === 192 && b === 0) ||
    (a === 192 && b === 0 && c === 2) ||
    (a === 198 && (b === 18 || b === 19 || b === 51)) ||
    (a === 203 && b === 0 && c === 113) ||
    a >= 224
  );
}

function isBlockedIPv6(address: string): boolean {
  const value = address.toLowerCase();
  if (value === "::" || value === "::1") return true;
  if (value.startsWith("::ffff:")) {
    return isBlockedIPv4(value.replace("::ffff:", ""));
  }

  const first = Number.parseInt(value.split(":")[0], 16);
  if (!Number.isFinite(first)) return true;

  return (
    (first & 0xfe00) === 0xfc00 ||
    (first & 0xffc0) === 0xfe80 ||
    (first & 0xff00) === 0xff00 ||
    value.startsWith("2001:db8:")
  );
}

function isRedirect(status: number): boolean {
  return status >= 300 && status < 400;
}

async function readLimitedArrayBuffer(resp: Response): Promise<ArrayBuffer> {
  const contentLength = Number(resp.headers.get("content-length") ?? "0");
  if (contentLength > MAX_PDF_SIZE) {
    throw new Error(
      `PDF too large: ${(contentLength / 1024 / 1024).toFixed(1)} MB (max ${MAX_PDF_SIZE / 1024 / 1024} MB)`,
    );
  }

  if (!resp.body) {
    const buffer = await resp.arrayBuffer();
    if (buffer.byteLength > MAX_PDF_SIZE) {
      throw new Error(
        `PDF too large: ${(buffer.byteLength / 1024 / 1024).toFixed(1)} MB (max ${MAX_PDF_SIZE / 1024 / 1024} MB)`,
      );
    }
    return buffer;
  }

  const reader = resp.body.getReader();
  const chunks: Uint8Array[] = [];
  let received = 0;

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    if (!value) continue;

    received += value.byteLength;
    if (received > MAX_PDF_SIZE) {
      await reader.cancel();
      throw new Error(
        `PDF too large: ${(received / 1024 / 1024).toFixed(1)} MB (max ${MAX_PDF_SIZE / 1024 / 1024} MB)`,
      );
    }
    chunks.push(value);
  }

  const merged = new Uint8Array(received);
  let offset = 0;
  for (const chunk of chunks) {
    merged.set(chunk, offset);
    offset += chunk.byteLength;
  }

  return merged.buffer;
}

export function splitIntoSections(
  text: string,
): Array<{ name: string; text: string }> {
  const lines = text.split("\n");
  const sections: Array<{ name: string; text: string }> = [];
  let currentName = "Full Text";
  let currentLines: string[] = [];

  for (const line of lines) {
    const stripped = line.trim();
    if (!stripped) {
      currentLines.push("");
      continue;
    }

    // Check if this line is a section header
    const matchedSection = matchSectionHeader(stripped);
    if (matchedSection) {
      // Save previous section
      if (currentLines.length > 0) {
        const sectionText = currentLines.join("\n").trim();
        if (sectionText) {
          sections.push({ name: currentName, text: sectionText });
        }
      }
      currentName = matchedSection;
      currentLines = [];
    } else {
      currentLines.push(line);
    }
  }

  // Save last section
  if (currentLines.length > 0) {
    const sectionText = currentLines.join("\n").trim();
    if (sectionText) {
      sections.push({ name: currentName, text: sectionText });
    }
  }

  // If no sections were detected (only "Full Text"), return as-is
  if (sections.length <= 1) {
    return [{ name: "Full Text", text: text.trim() }];
  }

  // Filter out unwanted sections
  const filtered = sections.filter((s) => {
    const nameLower = s.name.toLowerCase();
    if (EXCLUDE_SECTIONS.has(nameLower)) return false;
    if (s.text.length < 50) return false;
    return true;
  });

  return filtered.length > 0
    ? filtered
    : [{ name: "Full Text", text: text.trim() }];
}

function matchSectionHeader(line: string): string | null {
  for (const pattern of SECTION_PATTERNS) {
    const m = pattern.exec(line);
    if (m) {
      const rawName = m[1] ?? line;
      return normalizeSectionName(rawName);
    }
  }
  return null;
}

function normalizeSectionName(name: string): string {
  const nameLower = name.toLowerCase().trim();

  const mapping: Record<string, string> = {
    abstract: "Abstract",
    introduction: "Introduction",
    background: "Background",
    method: "Methods",
    methods: "Methods",
    "materials and methods": "Methods",
    "materials & methods": "Methods",
    "experimental methods": "Methods",
    result: "Results",
    results: "Results",
    "results and discussion": "Results and Discussion",
    discussion: "Discussion",
    conclusion: "Conclusion",
    conclusions: "Conclusion",
    "concluding remarks": "Conclusion",
    limitation: "Limitations",
    limitations: "Limitations",
    acknowledgement: "Acknowledgements",
    acknowledgements: "Acknowledgements",
    acknowledgment: "Acknowledgements",
    acknowledgments: "Acknowledgements",
    references: "References",
    bibliography: "References",
    supplementary: "Supplementary",
    "supporting information": "Supplementary",
  };

  return (
    mapping[nameLower] ??
    name.charAt(0).toUpperCase() + name.slice(1).toLowerCase()
  );
}
